{
    "active_lock_reason": null,
    "assignee": null,
    "assignees": [],
    "author_association": "CONTRIBUTOR",
    "body": "## Feedback or description of feature requested\r\n\r\nAfter upgrading from gmt `6.0.0_r20005` to `6.0.0_r20469` of upstream trunk gmt library, and every commit since,  `gmt-python` stopped functioning (import fails with CLIB Error). I understand these are development commits, I am maintaining some images with cutting edge builds of gmt for test purposes. I have done due diligence to verify it isn't other incompatabilities, or build errors. This is a change in behavior from previous versions of GMT which I have tested with development versions of `gmt-python`\r\n\r\nIt appears an API change may been introduced in [r20437](http://gmt.soest.hawaii.edu/projects/gmt/repository/revisions/20437/diff/trunk/src/gmt_api.c). An API Pointer has been added to the function signature, this may contribute to the error. GMT_PAD_DEFAULT was moved from a pre processor constant to enum in [Revision 18648](http://gmt.soest.hawaii.edu/projects/gmt/repository/revisions/18648) and gmt-python was updated to accommodate this change at some point\r\n\r\nBefore and after docker images are available. for verification  `nc5ng/gmt:6.0.0_r20005-python` (Functioning) and `nc5ng/gmt:6.0.0_r20469-python` (Broken).  [docker hub link](https://hub.docker.com/r/nc5ng/) \r\n\r\nnote: these have slightly different versions of `gmt-python`, but I have downgraded/upgraded to verify and it is not due to `gmt-python`.\r\n\r\n\r\n\r\n\r\n\r\n\r\n**Are you willing to help implement and maintain this bug?** Yes if guidance is provided\r\n\r\n\r\n## Full code that generated the error\r\n\r\n```python\r\nimport gmt\r\n```\r\n\r\n## Full error message\r\n```\r\n/usr/local/lib/python3.5/dist-packages/gmt/__init__.py in <module>()\r\n     26 \r\n     27 # Start our global modern mode session\r\n---> 28 _begin()\r\n     29 # Tell Python to run _end when shutting down\r\n     30 _atexit.register(_end)\r\n\r\n/usr/local/lib/python3.5/dist-packages/gmt/session_management.py in begin()\r\n     15     \"\"\"\r\n     16     prefix = 'gmt-python-session'\r\n---> 17     with LibGMT() as lib:\r\n     18         lib.call_module('begin', prefix)\r\n     19 \r\n\r\n/usr/local/lib/python3.5/dist-packages/gmt/clib/core.py in __enter__(self)\r\n    128         Start the GMT session and keep the session argument.\r\n    129         \"\"\"\r\n--> 130         self.current_session = self.create_session('gmt-python-session')\r\n    131         return self\r\n    132 \r\n\r\n/usr/local/lib/python3.5/dist-packages/gmt/clib/core.py in create_session(self, session_name)\r\n    170         # print function.\r\n    171         print_func = None\r\n--> 172         padding = self.get_constant('GMT_PAD_DEFAULT')\r\n    173         session_type = self.get_constant('GMT_SESSION_EXTERNAL')\r\n    174         session = c_create_session(session_name.encode(), padding,\r\n\r\n/usr/local/lib/python3.5/dist-packages/gmt/clib/core.py in get_constant(self, name)\r\n    237         if value is None or value == -99999:\r\n    238             raise GMTCLibError(\r\n--> 239                 \"Constant '{}' doesn't exits in libgmt.\".format(name))\r\n    240 \r\n    241         return value\r\n   GMTCLibError: Constant 'GMT_PAD_DEFAULT' doesn't exits in libgmt.\r\n```\r\n\r\n## System information\r\n\r\n* Operating system: Ubuntu 16.04.5 (Docker Image)\r\n* Python installation (Anaconda, system, ETS): [python3 from ubuntu repository](https://packages.ubuntu.com/xenial/python3)\r\n* Version of GMT: Tested `6.0.0_r20005` (works) and `6.0.0_r20471` (broken) and `6.0.0_r20469` (broken)\r\n* Version of Python: 3.5.1-3\r\n* Version of this package: `0.1a3`, `0.1a3+112.gefabb47` and `0.1a3+79.gf979f02'\r\n\r\n\r\n",
    "closed_at": "2018-08-17T04:06:37Z",
    "closed_by": {
        "avatar_url": "https://avatars.githubusercontent.com/u/290082?v=4",
        "events_url": "https://api.github.com/users/leouieda/events{/privacy}",
        "followers_url": "https://api.github.com/users/leouieda/followers",
        "following_url": "https://api.github.com/users/leouieda/following{/other_user}",
        "gists_url": "https://api.github.com/users/leouieda/gists{/gist_id}",
        "gravatar_id": "",
        "html_url": "https://github.com/leouieda",
        "id": 290082,
        "login": "leouieda",
        "node_id": "MDQ6VXNlcjI5MDA4Mg==",
        "organizations_url": "https://api.github.com/users/leouieda/orgs",
        "received_events_url": "https://api.github.com/users/leouieda/received_events",
        "repos_url": "https://api.github.com/users/leouieda/repos",
        "site_admin": false,
        "starred_url": "https://api.github.com/users/leouieda/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/leouieda/subscriptions",
        "type": "User",
        "url": "https://api.github.com/users/leouieda"
    },
    "comments": 5,
    "comments_content": [
        {
            "author_association": "NONE",
            "body": "\ud83d\udc4b Thanks for opening your first issue here! Please make sure you filled out the template with as much detail as possible. You might also want to take a look at our [contributing guidelines](https://github.com/GenericMappingTools/gmt-python/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/GenericMappingTools/gmt-python/blob/master/CODE_OF_CONDUCT.md).\n",
            "created_at": "2018-08-03T01:59:38Z",
            "html_url": "https://github.com/GenericMappingTools/pygmt/issues/222#issuecomment-410121109",
            "id": 410121109,
            "issue_url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/222",
            "node_id": "MDEyOklzc3VlQ29tbWVudDQxMDEyMTEwOQ==",
            "performed_via_github_app": null,
            "reactions": {
                "+1": 0,
                "-1": 0,
                "confused": 0,
                "eyes": 0,
                "heart": 0,
                "hooray": 0,
                "laugh": 0,
                "rocket": 0,
                "total_count": 0,
                "url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/comments/410121109/reactions"
            },
            "updated_at": "2018-08-03T01:59:38Z",
            "url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/comments/410121109",
            "user": {
                "avatar_url": "https://avatars.githubusercontent.com/in/4141?v=4",
                "events_url": "https://api.github.com/users/welcome%5Bbot%5D/events{/privacy}",
                "followers_url": "https://api.github.com/users/welcome%5Bbot%5D/followers",
                "following_url": "https://api.github.com/users/welcome%5Bbot%5D/following{/other_user}",
                "gists_url": "https://api.github.com/users/welcome%5Bbot%5D/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/apps/welcome",
                "id": 30606887,
                "login": "welcome[bot]",
                "node_id": "MDM6Qm90MzA2MDY4ODc=",
                "organizations_url": "https://api.github.com/users/welcome%5Bbot%5D/orgs",
                "received_events_url": "https://api.github.com/users/welcome%5Bbot%5D/received_events",
                "repos_url": "https://api.github.com/users/welcome%5Bbot%5D/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/welcome%5Bbot%5D/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/welcome%5Bbot%5D/subscriptions",
                "type": "Bot",
                "url": "https://api.github.com/users/welcome%5Bbot%5D"
            }
        },
        {
            "author_association": "CONTRIBUTOR",
            "body": "Confirmed by making a quick fix to [`session.py#L216-L218`]( https://github.com/GenericMappingTools/gmt-python/blob/master/gmt/clib/session.py#L216-L218)\r\n\r\n```python\r\nc_get_enum = self.get_libgmt_func(\r\n            \"GMT_Get_Enum\", argtypes=[ctp.c_char_p], restype=ctp.c_int\r\n        )\r\n\r\nvalue = c_get_enum(name.encode())\r\n```\r\n\r\nchanged to \r\n\r\n\r\n```python\r\nc_get_enum = self.get_libgmt_func(\r\n            \"GMT_Get_Enum\", argtypes=[ctp.c_void_p, ctp.c_char_p], restype=ctp.c_int\r\n        )\r\n\r\nvalue = c_get_enum(None, name.encode())\r\n```\r\n\r\nThis works because API change accepts a null pointer for new argument\r\n\r\nCommit on my fork here: https://github.com/nc5ng/gmt-python/commit/c539b627305756431acdef73ff266522c96b8880\r\n\r\n\r\nVerified on the `gmt 6.0.0_r20471` and `gmt-python 0.1a3+113.gc539b62`",
            "created_at": "2018-08-03T02:15:36Z",
            "html_url": "https://github.com/GenericMappingTools/pygmt/issues/222#issuecomment-410123482",
            "id": 410123482,
            "issue_url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/222",
            "node_id": "MDEyOklzc3VlQ29tbWVudDQxMDEyMzQ4Mg==",
            "performed_via_github_app": null,
            "reactions": {
                "+1": 0,
                "-1": 0,
                "confused": 0,
                "eyes": 0,
                "heart": 0,
                "hooray": 0,
                "laugh": 0,
                "rocket": 0,
                "total_count": 0,
                "url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/comments/410123482/reactions"
            },
            "updated_at": "2018-08-03T02:15:36Z",
            "url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/comments/410123482",
            "user": {
                "avatar_url": "https://avatars.githubusercontent.com/u/6558063?v=4",
                "events_url": "https://api.github.com/users/akshmakov/events{/privacy}",
                "followers_url": "https://api.github.com/users/akshmakov/followers",
                "following_url": "https://api.github.com/users/akshmakov/following{/other_user}",
                "gists_url": "https://api.github.com/users/akshmakov/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/akshmakov",
                "id": 6558063,
                "login": "akshmakov",
                "node_id": "MDQ6VXNlcjY1NTgwNjM=",
                "organizations_url": "https://api.github.com/users/akshmakov/orgs",
                "received_events_url": "https://api.github.com/users/akshmakov/received_events",
                "repos_url": "https://api.github.com/users/akshmakov/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/akshmakov/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/akshmakov/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/akshmakov"
            }
        },
        {
            "author_association": "CONTRIBUTOR",
            "body": "Possibly related to #215",
            "created_at": "2018-08-03T03:28:17Z",
            "html_url": "https://github.com/GenericMappingTools/pygmt/issues/222#issuecomment-410133900",
            "id": 410133900,
            "issue_url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/222",
            "node_id": "MDEyOklzc3VlQ29tbWVudDQxMDEzMzkwMA==",
            "performed_via_github_app": null,
            "reactions": {
                "+1": 0,
                "-1": 0,
                "confused": 0,
                "eyes": 0,
                "heart": 0,
                "hooray": 0,
                "laugh": 0,
                "rocket": 0,
                "total_count": 0,
                "url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/comments/410133900/reactions"
            },
            "updated_at": "2018-08-03T03:28:17Z",
            "url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/comments/410133900",
            "user": {
                "avatar_url": "https://avatars.githubusercontent.com/u/6558063?v=4",
                "events_url": "https://api.github.com/users/akshmakov/events{/privacy}",
                "followers_url": "https://api.github.com/users/akshmakov/followers",
                "following_url": "https://api.github.com/users/akshmakov/following{/other_user}",
                "gists_url": "https://api.github.com/users/akshmakov/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/akshmakov",
                "id": 6558063,
                "login": "akshmakov",
                "node_id": "MDQ6VXNlcjY1NTgwNjM=",
                "organizations_url": "https://api.github.com/users/akshmakov/orgs",
                "received_events_url": "https://api.github.com/users/akshmakov/received_events",
                "repos_url": "https://api.github.com/users/akshmakov/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/akshmakov/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/akshmakov/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/akshmakov"
            }
        },
        {
            "author_association": "MEMBER",
            "body": "Hi @akshmakov, thanks for the detailed report and investigating the issue! I really appreciate it!\r\n\r\nThis was a recent change to the API. `GMT_Get_Enum` was one of the only functions that didn't take an `API` pointer as the first argument. This was recently changed to make the API more uniform and allow for possible extensions in the future. Your change should probably work (great investigating, BTW).\r\n\r\nWould you like to open a pull request with that fix? Let me know if you need any guidance for this.\r\n\r\nYour build will break for now because of #221 but once I have that fixed everything should be fine.\r\n\r\nAgain, big thanks for investigating this! :1st_place_medal: ",
            "created_at": "2018-08-03T18:26:25Z",
            "html_url": "https://github.com/GenericMappingTools/pygmt/issues/222#issuecomment-410337991",
            "id": 410337991,
            "issue_url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/222",
            "node_id": "MDEyOklzc3VlQ29tbWVudDQxMDMzNzk5MQ==",
            "performed_via_github_app": null,
            "reactions": {
                "+1": 0,
                "-1": 0,
                "confused": 0,
                "eyes": 0,
                "heart": 0,
                "hooray": 0,
                "laugh": 0,
                "rocket": 0,
                "total_count": 0,
                "url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/comments/410337991/reactions"
            },
            "updated_at": "2018-08-03T18:26:25Z",
            "url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/comments/410337991",
            "user": {
                "avatar_url": "https://avatars.githubusercontent.com/u/290082?v=4",
                "events_url": "https://api.github.com/users/leouieda/events{/privacy}",
                "followers_url": "https://api.github.com/users/leouieda/followers",
                "following_url": "https://api.github.com/users/leouieda/following{/other_user}",
                "gists_url": "https://api.github.com/users/leouieda/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/leouieda",
                "id": 290082,
                "login": "leouieda",
                "node_id": "MDQ6VXNlcjI5MDA4Mg==",
                "organizations_url": "https://api.github.com/users/leouieda/orgs",
                "received_events_url": "https://api.github.com/users/leouieda/received_events",
                "repos_url": "https://api.github.com/users/leouieda/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/leouieda/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/leouieda/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/leouieda"
            }
        },
        {
            "author_association": "CONTRIBUTOR",
            "body": "@leouieda no problem I will reword  the commit to reference gmt commit and submit a pr today or tomorrow \r\n\r\nA few questions\r\n\r\n1. This is a distinct break with previous interface, so will require gmt 6.0.0 after the specific revision with this change. Would some kind of test or check for gmt version be preferable as an intermediate step to support both versions?\r\n\r\n2. Are you handling API pointers in other calls that I should copy? Is passing a null void pointer sufficient?\r\n\r\n\r\nThis does build on my machine with `pip install` directly from repo, is #221 related to your release  build system?\r\n",
            "created_at": "2018-08-04T01:19:54Z",
            "html_url": "https://github.com/GenericMappingTools/pygmt/issues/222#issuecomment-410411842",
            "id": 410411842,
            "issue_url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/222",
            "node_id": "MDEyOklzc3VlQ29tbWVudDQxMDQxMTg0Mg==",
            "performed_via_github_app": null,
            "reactions": {
                "+1": 0,
                "-1": 0,
                "confused": 0,
                "eyes": 0,
                "heart": 0,
                "hooray": 0,
                "laugh": 0,
                "rocket": 0,
                "total_count": 0,
                "url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/comments/410411842/reactions"
            },
            "updated_at": "2018-08-04T01:19:54Z",
            "url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/comments/410411842",
            "user": {
                "avatar_url": "https://avatars.githubusercontent.com/u/6558063?v=4",
                "events_url": "https://api.github.com/users/akshmakov/events{/privacy}",
                "followers_url": "https://api.github.com/users/akshmakov/followers",
                "following_url": "https://api.github.com/users/akshmakov/following{/other_user}",
                "gists_url": "https://api.github.com/users/akshmakov/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/akshmakov",
                "id": 6558063,
                "login": "akshmakov",
                "node_id": "MDQ6VXNlcjY1NTgwNjM=",
                "organizations_url": "https://api.github.com/users/akshmakov/orgs",
                "received_events_url": "https://api.github.com/users/akshmakov/received_events",
                "repos_url": "https://api.github.com/users/akshmakov/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/akshmakov/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/akshmakov/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/akshmakov"
            }
        }
    ],
    "comments_url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/222/comments",
    "created_at": "2018-08-03T01:59:35Z",
    "events_url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/222/events",
    "html_url": "https://github.com/GenericMappingTools/pygmt/issues/222",
    "id": 347232394,
    "labels": [
        {
            "color": "ee0701",
            "default": true,
            "description": "Something isn't working",
            "id": 563309827,
            "name": "bug",
            "node_id": "MDU6TGFiZWw1NjMzMDk4Mjc=",
            "url": "https://api.github.com/repos/GenericMappingTools/pygmt/labels/bug"
        }
    ],
    "labels_url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/222/labels{/name}",
    "locked": false,
    "milestone": null,
    "node_id": "MDU6SXNzdWUzNDcyMzIzOTQ=",
    "number": 222,
    "performed_via_github_app": null,
    "reactions": {
        "+1": 0,
        "-1": 0,
        "confused": 0,
        "eyes": 0,
        "heart": 0,
        "hooray": 0,
        "laugh": 0,
        "rocket": 0,
        "total_count": 0,
        "url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/222/reactions"
    },
    "repository_url": "https://api.github.com/repos/GenericMappingTools/pygmt",
    "state": "closed",
    "state_reason": "completed",
    "timeline_url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/222/timeline",
    "title": "GMT_PAD_DEFAULT doesn't exist in libgmt:  Possible upstream GMT_GET_Enum API Change ",
    "updated_at": "2018-08-17T04:06:37Z",
    "url": "https://api.github.com/repos/GenericMappingTools/pygmt/issues/222",
    "user": {
        "avatar_url": "https://avatars.githubusercontent.com/u/6558063?v=4",
        "events_url": "https://api.github.com/users/akshmakov/events{/privacy}",
        "followers_url": "https://api.github.com/users/akshmakov/followers",
        "following_url": "https://api.github.com/users/akshmakov/following{/other_user}",
        "gists_url": "https://api.github.com/users/akshmakov/gists{/gist_id}",
        "gravatar_id": "",
        "html_url": "https://github.com/akshmakov",
        "id": 6558063,
        "login": "akshmakov",
        "node_id": "MDQ6VXNlcjY1NTgwNjM=",
        "organizations_url": "https://api.github.com/users/akshmakov/orgs",
        "received_events_url": "https://api.github.com/users/akshmakov/received_events",
        "repos_url": "https://api.github.com/users/akshmakov/repos",
        "site_admin": false,
        "starred_url": "https://api.github.com/users/akshmakov/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/akshmakov/subscriptions",
        "type": "User",
        "url": "https://api.github.com/users/akshmakov"
    }
}